### DISCUSSION TITLE
How to programmatically authenticate to URL returned by asExternalUri with Remote Tunnels?

### DISCUSSION THREAD

**Discussion by nisargjhaveri** [microsoft/vscode-discussions#1528] [Extension Development QnA] [answered]
I'm developing an [extension to access local Android device on the remote machine](https://github.com/nisargjhaveri/vscode-remote-adb-server) when using VS Code Remote Development. For this I'm planning to have two extensions. One running on the remote extension host, which will create a http server, and another running on the UI extension host, which will connect to the local devices and forward communication to the remote http server. 

I'm using `asExternalUri` to forward the port and connect to the remote host from the UI counterpart. This works well with Remote SSH, but I'm having some trouble when using this with Remote Tunnels.

When using `asExternalUri` with Remote Tunnels, it generates a URL which requires additional authentication. How do I authenticate to this URL programmatically from within the same VS Code remote session?

**Comment by robconery** [comment]
Can you be a bit more specific about the issue? What is the format of the URL, and what additional information is needed? If I understand correctly, your remote http service is what will be authenticating the URL (provided you have port access, etc), so knowing more about the issue would be helpful.

**Reply by nisargjhaveri** [reply]
I'm trying to develop a solution for sharing local Android devices to the remote host for development. For this, I have two extensions. One "remote" extension, and another "local" extension.

The "remote" extension starts a server and shares the url to that server via `asExternalUri`, which handles the port forwarding and makes the server available on local.

The "local" extension connects to this URL to provide the relevant functionality.

All these works very well with "Remote - SSH", but does not work as intended with "Remote - Tunnels". The URL returned by `asExternalUri` in case of tunnels with `https://*.devtunnels.ms/` format requires authentication to connect to and the "local" extension by itself cannot easily authenticate to this URL programmatically.

Is there a way to get access token to this URL when running in the same VSCode instance, or obtaining pre-authenticated URL since the extension is not supposed to make any assumptions about the URL returned by `asExternalUri` as mentioned in the docs.

**Reply by robconery** [reply]
Thanks for the additional info. @isidorn might have some input on this.

**Reply by isidorn** [reply]
I do not know. Maybe @connor4312 might be able to help.

**Comment by burkeholland** [comment]
@nisargjhaveri Is this still an issue?

**Reply by nisargjhaveri** [reply]
Yes, I'm still looking for a solution to this, I would really appreciate if someone could help. Several recommendations in https://code.visualstudio.com/api/advanced-topics/remote-extensions are not working out well with Remote Tunnels due to such authentication issues.

**Comment by connor4312** [comment, marked as answer]
Covered by https://github.com/microsoft/vscode-remote-release/issues/9195 which I will take up soon
